x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+
+Continuation of [`x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts`](x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts):
+
+```ts
+} else if (rule != null && overwriteRules) {
+
+  /* Perform migration logic on rules params before patching: if neither the rule 
+   * parsed from the ndjson input nor the existent rule saved object have a
+   * `prebuilt` property, set the `prebuilt` field to `immutable` value set
+   * in the input field, or as fallback, as set in the existing rule.
+   * The customized field should take the value set in the input, or default to false.
+   */
+  let migratedParams;
+  if (parsedRule.prebuilt === undefined && rule.params.prebuilt === undefined) {
+      migratedParams = {
+          ...parsedRule,
+          prebuilt: parsedRule.immutable || rule.params.immutable,
+          customized: parsedRule?.customized ?? false,
+      };
+  }
+
+  await patchRules({
+    rulesClient,
+    existingRule: rule,
+    nextParams: {
+      ...migratedParams,
+      exceptions_list: [...exceptions],
+    },
+    allowMissingConnectorSecrets,
+    shouldIncrementRevision: false,
+  });
+  resolve({
+    rule_id: parsedRule.rule_id,
+    status_code: 200,
+  });
+}
+```
+
+$\space$
+
+
+## Exporting and importing rules
+
+**Current behaviour:**
+
+- Custom Rules can be exported and imported
+- When trying to import a custom rule, its `rule_id` should not already exist in any of the installed rules. Otherwise the import logic will fail : `rule_id: {rule_id} already exists`
+- Prebuilt Rules cannot be exported
+- Prebuilt Rules cannot be imported (if you try to import a rule that has been set the `immutable` field to `true` manually, the import endpoint will reject it as the `immutable` field is expected to be `false`)
+
+**Updated behaviour**
+
+We will now allow the user to export and import both **prebuilt** and **custom rules**.
+
+### Exporting rules
+
+The user will now be able to export both custom and prebuilt rules (including any customizations that might have been performed on its modifiable fields).
+
+The export endpoints (bulk export and bulk action in export mode) will not carry out any migration logic and will export the rule as-is. Since these endpoints only read the rules and create an `ndjson` file to make them available for export, we don't want to add any additional writing/patching logic to migrate the rule's schema on Elasticsearch.
```

Comment:
> The export endpoints (bulk export and bulk action in export mode) will not carry out any migration logic and will export the rule as-is

Can you elaborate why? I think the export endpoint should normalize rule saved objects to the new schema ("migrate on-read") before returning them from the export endpoints, just like any other "GET" endpoints that return rules.

Might be that there's a confusion here around physical migration on-write (with saving changes to ES) vs virtual normalization on-read (w/o saving changes to ES). Commented on this above in some comment. Let's clearly distinguish these two types of transformations in the RFC and use different terms for them.